src/server/pfs/server/master.go

```diff
@@ -70,21 +86,32 @@ func (d *driver) master(ctx context.Context) {
 }
 
 func (d *driver) finishCommits(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	ring, err := consistenthashing.New(ctx, path.Join(d.prefix, masterLockPath, "ring"), d.etcdClient, d.log)
+	defer cancel()
+	if err != nil {
+		return err
+	}
 	repos := make(map[string]context.CancelFunc)
 	defer func() {
 		for _, cancel := range repos {
 			cancel()
 		}
 	}()
 	compactor := newCompactor(d.storage, d.env.StorageConfig.StorageCompactionMaxFanIn)
-	err := d.repos.ReadOnly(ctx).WatchF(func(ev *watch.Event) error {
+	err = d.repos.ReadOnly(ctx).WatchF(func(ev *watch.Event) error {
 		if ev.Type == watch.EventError {
 			return ev.Err
 		}
 		key := string(ev.Key)
+		lockPrefix := path.Join(d.prefix, masterLockPath, "repos", key)
 		if ev.Type == watch.EventDelete {
 			if cancel, ok := repos[key]; ok {
+				err = ring.Unlock(lockPrefix)
```

Comment:
We can just return without the cancel if err isn't nil. All of the repos will get canceled in that case.